Annotate disassembly with register‐resident variable locations

UltimateForce21 opened the pull request:
This PR builds upon the functionality introduced in [PR #144238](https://github.com/llvm/llvm-project/pull/144238), which adds support for annotating disassembly output with DWARF variable locations.

This patch includes:

Additional test coverage for various DWARF location cases (e.g., stack variables, control flow edges, partial decoding).

A refined implementation that uses the new DWARFExpressionEntry API.

Logic to gracefully skip <decoding error> fragments while preserving valid annotations.

⚠️ Note: This PR depends on #144238 and should be rebased once it lands.

### Summary
This patch adds a first‐pass prototype of “rich disassembly” annotations into Instruction::Dump(). For each instruction, it will:

1. Query our new `GetExpressionEntryAtAddress()` API for any in‐scope variable live‐ranges covering this PC.

If a variable’s live‐range begins exactly at the current PC that is being iterated in the disassembly, call `expr->DumpLocation(...)` to render its DWARF expression (e.g. “`DW_OP_reg5 RDI`”, “`DW_OP_reg3 RBX`”, or constant) and append it as a `; var = …` comment after the instruction.

Currently this only handles simple cases:

Variables whose location expression is a single register (e.g. `DW_OP_reg14 R14`)

Variables whose location expression is a constant (e.g. `DW_OP_consts +0`)

### Example test program
```
// test.cpp
#include <iostream>

void testFunction(int value);

int main(int argc, char** argv) {
    int total = 0;
    for (int my_special_variable = 0; my_special_variable < 5; ++my_special_variable){
        puts(argv[my_special_variable]);
        int new_variable = my_special_variable + 1;
        testFunction(new_variable); 
        total += my_special_variable;
    }
    return total;
}


void testFunction(int value) {
    std::cout << "Local variable value: " << value << std::endl;
}
```


### Disassembly output with annotations

```
(lldb) target create "/home/ultimateforce21/lldb_test/test"
Current executable set to '/home/ultimateforce21/lldb_test/test' (x86_64).
(lldb) b main
Breakpoint 1: where = test`main + 16 at test.cpp:10:14, address = 0x0000000000001200
(lldb) r
Process 26674 launched: '/home/ultimateforce21/lldb_test/test' (x86_64)
Process 26674 stopped
* thread #1, name = 'test', stop reason = breakpoint 1.1
    frame #0: 0x0000555555555200 test`main(argc=<unavailable>, argv=0x00007fffffffdc68) at test.cpp:10:14
   7    int main(int argc, char** argv) {
   8        int total = 0;
   9        for (int my_special_variable = 0; my_special_variable < 5; ++my_special_variable){
-> 10           puts(argv[my_special_variable]);
   11           int new_variable = my_special_variable + 1;
   12           testFunction(new_variable); 
   13           total += my_special_variable;
(lldb) dis
test`main:
    0x5555555551f0 <+0>:  pushq  %r14           ; argc = DW_OP_reg5 RDI ; argv = DW_OP_reg4 RSI ; std::__ioinit = DW_OP_addrx 0x0
    0x5555555551f2 <+2>:  pushq  %rbx
    0x5555555551f3 <+3>:  pushq  %rax
    0x5555555551f4 <+4>:  movq   %rsi, %r14
    0x5555555551f7 <+7>:  xorl   %ebx, %ebx ; total = DW_OP_consts +0, DW_OP_stack_value
    0x5555555551f9 <+9>:  nopl   (%rax) ; my_special_variable = DW_OP_reg3 RBX ; argc = DW_OP_entry_value(DW_OP_reg5 RDI), DW_OP_stack_value ; argv = DW_OP_reg14 R14
->  0x555555555200 <+16>: movq   (%r14,%rbx,8), %rdi
    0x555555555204 <+20>: callq  0x5555555550a0 ; symbol stub for: puts
    0x555555555209 <+25>: addq   $0x1, %rbx
    0x55555555520d <+29>: movl   %ebx, %edi ; new_variable = DW_OP_reg3 RBX
    0x55555555520f <+31>: callq  0x555555555230 ; testFunction at test.cpp:19
    0x555555555214 <+36>: cmpq   $0x5, %rbx ; my_special_variable = DW_OP_reg3 RBX
    0x555555555218 <+40>: jne    0x555555555200 ; <+16> at test.cpp:10:14
    0x55555555521a <+42>: movl   $0xa, %eax
    0x55555555521f <+47>: addq   $0x8, %rsp
    0x555555555223 <+51>: popq   %rbx
    0x555555555224 <+52>: popq   %r14
    0x555555555226 <+54>: retq                             ; argv = DW_OP_entry_value(DW_OP_reg4 RSI), DW_OP_stack_value
```




### Next steps
**Cleanup & style (DONE)**
- Tidy up any styling according to LLDB’s conventions, and handle any off‐by‐one column issues.

**Test coverage (DONE)**
- Add a lldb-test cases exercising single-register variables at known instructions.

UltimateForce21 committed:
[lldb] Add DWARFExpressionEntry and GetExpressionEntryAtAddress() to DWARFExpressionList

This introduces a new API for retrieving DWARF expression metadata associated with variable location entries at a given PC address. It provides the base, end, and expression pointer for downstream consumers such as disassembler annotations.

Intended for use in richer instruction annotations in Instruction::Dump().

UltimateForce21 committed:
Update lldb/include/lldb/Expression/DWARFExpressionList.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

UltimateForce21 committed:
Update lldb/include/lldb/Expression/DWARFExpressionList.h

Updated comment for GetExpressionEntryAtAddress to directly refer to struct DWARFExpressionEntry

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

UltimateForce21 committed:
Update lldb/source/Expression/DWARFExpressionList.cpp

updating code style for function GetExpressionEntryAtAddress

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

UltimateForce21 committed:
Update DWARFExpressionList.h

Replace raw base/end with `AddressRange` in `DWARFExpressionEntry` and cleans up helper comments to follow Doxygen convention.

Using `AddressRange` makes the intent clearer, avoids duplication of basic `AddressRange` logic usage

UltimateForce21 committed:
Update DWARFExpressionList.cpp

Converts `GetExpressionEntryAtAddress` to return `llvm::Expected<DWARFExpressionEntry>` using the updated `DWARFExpressionEntry`. Updates the implementation to compute a single `AddressRange file_range` for each DWARF location interval.

UltimateForce21 committed:
Update DWARFExpressionList.h

Updated commenting style for struct DWARFExpressionEntry

UltimateForce21 committed:
Update DWARFExpressionList.cpp

Updated function `llvm::Expected<DWARFExpressionList::DWARFExpressionEntry>
DWARFExpressionList::GetExpressionEntryAtAddress` to use `FindEntryThatContains` instead of `FindEntryIndexThatContains`

UltimateForce21 committed:
Change GetExpressionEntryAtAddress to return std::optional instead of llvm::Expected

UltimateForce21 committed:
Update DWARFExpressionList.cpp

Co-authored-by: Adrian Prantl <adrian.prantl@gmail.com>

UltimateForce21 committed:
Add underflow/overflow checks to GetExpressionEntryAtAddressi
This patch adds explicit checks:
  - ensure `load_addr >= func_load_addr` to avoid underflow,
  - compute and verify a temporary delta variable, then verify `delta + m_func_file_addr` does
    not exceed `addr_t` max to avoid overflow.

UltimateForce21 committed:
Annotate Instruction::Dump() with DWARF variable locations

UltimateForce21 committed:
Added Initial Basic API test for rich variable annotation in disassembly output. Right now just checks if DW annotations show up for a basic program and that a variable location is annotated (i.e 'a = DW_OP_reg...').

UltimateForce21 committed:
Improved DWARF variable annotation printing and alignment

- Fixed an issue where variable location annotations were not shown if the current instruction address did not exactly match the DWARF base address. Now, annotations are shown as long as the PC is within the valid range.
- Improved alignment of annotation comments in Instruction::Dump(). While `FillLastLineToColumn` can sometimes overcompensate due to internal formatting or byte-width mismatches, the overall alignment is now significantly more consistent and readable.

UltimateForce21 committed:
Filter out partial DWARF decoding errors from disassembly annotations

Previously, when a DWARF expression contained any decoding error,
the entire variable location annotation was printed with the error,
e.g. `c = DW_OP_addr 0x0, <decoding error> 00 00 00`. This was
misleading and cluttered the disassembly view.

This patch improves the formatting by stripping out the
`<decoding error ...>` fragments while preserving the valid portions
of the expression, so that partial information like
`c = DW_OP_addr 0x0` can still be shown.

This allows the rich disassembler to give more meaningful variable
annotations, especially in optimized (-O1/-O2) builds where partial
DWARF corruption or unsupported expressions may occur.

UltimateForce21 committed:
Ignore annotations with only decoding errors

Handled edge case where the entire DWARF expression is a `<decoding error>`, ensuring no misleading or empty annotations are printed for such variables.

UltimateForce21 committed:
Add tests for disassembly variable annotations and decoding edge cases

This patch adds API tests to verify that DWARF variable location
annotations are correctly displayed in the disassembly output.

The tests cover:
- Local variables in loops and functions
- Multiple stack variables
- Control flow edge cases
- Different optimization levels (-O0, -O1, -O2)
- Ensuring decoding errors are excluded from output

UltimateForce21 committed:
Make file_range optional in DWARFExpressionEntry for always-valid expr

UltimateForce21 committed:
Rebase disassembler annotations branch onto updated DWARFExpressionEntry API

This rebases the `add-disassembler-annotations` work onto the
latest `add-dwarfexprentry-api` branch so that the instruction
annotation patches sit cleanly atop the new DWARFExpressionEntry
struct and helper API. All conflicts have been resolved and the
annotation code now integrates with the updated std::optional<AddressRange>-based
GetExpressionEntryAtAddress signature.

UltimateForce21 committed:
Add `PrintRegisterOnly` flag in `struct DIDumpOptions` and created new `DWARFExpression::DumpLocationWithOptions` for simplified expression printing

This patch introduces a PrintRegisterOnly flag to the DIDumpOptions struct, enabling concise rendering of DWARF expressions for disassembler annotations.

Key changes:
- Added DumpLocationWithOptions to DWARFExpression for flexible dumping with DIDumpOptions.
- Updated DWARFExpression::print and Operation::print to respect PrintRegisterOnly, rendering registers like RDI without DW_OP_ or llvm: prefixes.
- Suppressed <decoding error> output when PrintRegisterOnly is true to avoid clutter during register-only disassembly output.

These changes are motivated by LLDB’s rich disassembler feature, where annotations should match user-facing register names without DWARF-level detail.

Test impact:
Some rich-disassembler tests that relied on DW_OP_ for validation were deprecated. Updated tests aligned with the new formatting will be added next.

adrian-prantl commented:
As high-level comments:
- I would filter out the entry values, since this feature is meant to help a user understand the disassembly, and the entry values don't really exist in the function.
- It would help readability if all annotations started at the same column
- in a follow-up patch it would be great to outline the valid range of each expression
- is it possible to hide the DW_OP... part and just show the human-readable register name?
- `DW_OP_consts +0, DW_OP_stack_value` -> I would filter out anything that isn't a pure register. Then we can later recognize certain kinds of expression, for example, this is a constant, so it should just print `= 0`

jasonmolenda commented:
Incidentally, this is always something get a little confused by when reading assembly, we see it with backtraces.  A CALL instruction executes and the return address is the next instruction.  That confuses people looking at the assembly language -- the "pc" pointer is pointing to the next instruction! -- but we cheat it for source symbolication, subtracting 1 from the return address when we do source line lookup so we show the function call as "currently executing" when we really have the return address which may point to the next source line.  We have to do the same thing with dwarf location list lookups, where we might have a `noreturn` function call, and the variables available at the return address (which is the target of a branch from another part of the function) are drastically different.

I'm aways unsure of how much to lie to the user in these scenarios.  The general call is "at source level lie, at assembly level tell the truth of what the return address shows us".  It gets a little messier on a function that ends in a noreturn CALL as the last instruction, we don't want to show the return address as the next function in memory, even though that's what it really shows us.

UltimateForce21 committed:
Add high-level comment explaining rich disassembly annotation logic in Instruction::Dump

UltimateForce21 committed:
Add comment clarifying annotation column length check in Instruction::Dump

UltimateForce21 commented on lldb/source/Core/Disassembler.cpp:
```diff
@@ -702,6 +705,78 @@ void Instruction::Dump(lldb_private::Stream *s, uint32_t max_opcode_byte_size,
+          func_load_addr = sc.function->GetAddress().GetLoadAddress(target_sp.get());
+
+        
+        if(ss.GetSizeOfLastLine() < annotation_column) {
```
Yes, exactly. I have now added the comment in commit: [6e17f77c4300654c7664b52efc045983bb59721d](https://github.com/llvm/llvm-project/pull/147460/commits/6e17f77c4300654c7664b52efc045983bb59721d)

UltimateForce21 committed:
Refactor variable annotation logic in `Instruction::Dump` using `annotate_variable` lambda function

UltimateForce21 commented on lldb/source/Core/Disassembler.cpp:
```diff
@@ -702,6 +705,78 @@ void Instruction::Dump(lldb_private::Stream *s, uint32_t max_opcode_byte_size,
+
+          std::vector<std::string> annotations;
+
+          if (var_list_sp) {
```
Thanks for the suggestion! I've pushed the change in [31431c0](https://github.com/llvm/llvm-project/pull/147460/commits/31431c0c1e8348dec7761dc7052ca355bfc950dd), which moves the annotation logic into a local annotate_variables lambda with early exits.

UltimateForce21 committed:
Use range-based for loop for variable list iteration in Instruction::Dump

UltimateForce21 commented on lldb/source/Core/Disassembler.cpp:
```diff
@@ -702,6 +705,78 @@ void Instruction::Dump(lldb_private::Stream *s, uint32_t max_opcode_byte_size,
+            if (!annotations.empty()) {
+              ss.FillLastLineToColumn(annotation_column, ' ');
+              ss.PutCString(" ; ");
+              for (size_t i = 0; i < annotations.size(); ++i) {
```
Thanks for the suggestion! I've updated the loop to use a range-based for as recommended. The change is now committed in [9c5cb8f](https://github.com/llvm/llvm-project/pull/147460/commits/9c5cb8fa6d622cdd12db4c6a291265a991d3eb5e).

UltimateForce21 replied:
I replaced the manual string concatenation with `llvm::join` as suggested in commit 
[ffefe5f](https://github.com/llvm/llvm-project/pull/147460/commits/ffefe5f08ba1f62c8ebea4e28f421e50aa193f66).

UltimateForce21 commented on lldb/source/Core/Disassembler.cpp:
```diff
@@ -702,6 +705,78 @@ void Instruction::Dump(lldb_private::Stream *s, uint32_t max_opcode_byte_size,
+          std::vector<std::string> annotations;
+
+          if (var_list_sp) {
+            for (size_t i = 0; i < var_list_sp->GetSize(); ++i) {
```
Thanks for the suggestion! I've updated the loop to use a range-based for as recommended. The change is now committed in [9c5cb8f](https://github.com/llvm/llvm-project/pull/147460/commits/9c5cb8fa6d622cdd12db4c6a291265a991d3eb5e).

UltimateForce21 commented on lldb/source/Expression/DWARFExpressionList.cpp:
```diff
@@ -53,6 +54,37 @@ bool DWARFExpressionList::ContainsAddress(lldb::addr_t func_load_addr,
+
+  // Guard against overflow.
+  lldb::addr_t delta = load_addr - func_load_addr;
+  if (delta > std::numeric_limits<lldb::addr_t>::max() - m_func_file_addr)
```
From when I dug into MathExtras last time, I found that its `AddOverflow` and `SubOverflow` helpers are only defined for **signed types**.

I they rely on two’s-complement signed overflow semantics and return the truncated signed result. Since `lldb::addr_t` is an **unsigned address,** I think using those helpers would either fail to compile or silently convert to signed semantics. 

That's why I tried to implement the signed overflow check into the function.

UltimateForce21 committed:
Consolidated DumpLocation and DumpLocationWithOptions using default DIDumpOptions instead of having to introduce new function

UltimateForce21 commented on lldb/include/lldb/Expression/DWARFExpression.h:
```diff
@@ -158,6 +158,7 @@ class DWARFExpression {
   }
 
   void DumpLocation(Stream *s, lldb::DescriptionLevel level, ABI *abi) const;
+  void DumpLocationWithOptions(Stream *s, lldb::DescriptionLevel level, ABI *abi, llvm::DIDumpOptions options) const;
```
I've updated the code to use a single `DumpLocation` method with a default `DIDumpOptions` argument, as recommended. The `DumpLocationWithOptions` variant has been removed in commit [ca8510c](https://github.com/llvm/llvm-project/pull/147460/commits/ca8510c41add8d34a03eef9f8fce57fd42759b79).

UltimateForce21 committed:
Use `llvm::join` to simplify annotation output formatting

UltimateForce21 committed:
Merge branch 'main' into add-disassembler-annotations
Resolve conflicts in DWARFExpression files after upstream refactor

UltimateForce21 committed:
Fix formatting to match LLVM style

UltimateForce21 committed:
More formatting fixes

UltimateForce21 committed:
Fix formatting for code and tests

UltimateForce21 commented:
Hi @adrian-prantl,

While debugging the current CI test failures, I found that a number of existing LLDB API tests (e.g., `TestFrameSelect`, `TestInferiorAssert`) depend on the exact disassembly output.
Since my changes add new annotations in `Instruction::Dump`, they alter the default output and cause these tests to fail.

To address this, I’m considering gating the annotation output behind a new LLDB setting, for example:

```
settings set enable-rich-disassembly true
```
This would default to false, so the default disassembly output would be unchanged for tests and existing users, but developers/users could explicitly enable the richer annotated disassembly when desired.

My thinking is:

- CI and default workflows see the same output as before (tests continue to pass).
- Rich annotations are still available when activated.

Does that sound like a reasonable approach or is there a better/easier solution?

adrian-prantl commented:
> Hi @adrian-prantl,
> 
> While debugging the current CI test failures, I found that a number of existing LLDB API tests (e.g., `TestFrameSelect`, `TestInferiorAssert`) depend on the exact disassembly output. Since my changes add new annotations in `Instruction::Dump`, they alter the default output and cause these tests to fail.
> 
> To address this, I’m considering gating the annotation output behind a new LLDB setting, for example:
> 
> ```
> settings set enable-rich-disassembly true
> ```

This, or adding an option to the dissemble CommandObject is the right solution. Personally I think I would prefer an option to the CommandObject. Do you think this is feasible, or are there more many points through which the disassembly is reachable, making an option infeasible.

> This would default to false, so the default disassembly output would be unchanged for tests and existing users, but developers/users could explicitly enable the richer annotated disassembly when desired.

I think that is correct while the feature is under development, but I think the goal should be for it to be on by default, with the few tests relying on it not being there opting out.

> My thinking is:
> 
> * CI and default workflows see the same output as before (tests continue to pass).
> * Rich annotations are still available when activated.
> 
> Does that sound like a reasonable approach or is there a better/easier solution?

Yes, this is the right approach!

UltimateForce21 commented:
Thanks, that makes sense.

Given what I found in the failing tests (some go through `SBFrame::Disassemble()`, others through the disassemble CLI), I see two viable rollouts:

### Phase 1 (minimal blast radius):

- Thread an `enable_rich` boolean into the disassembly path.

- Pass false from both `SBFrame::Disassemble()` and `CommandObjectDisassemble` by default, so current output stays identical and CI/IDEs remain unaffected.

- Add the CLI override (e.g. `disassemble --rich`) that sets `enable_rich=true` for that invocation.


### Phase 2 (once we’re confident):

- Flip the default to on in the CLI path (and/or provide an SB overload later), and update the handful of tests that rely on the old format to opt out (e.g. `disassemble --no-rich`). This matches our long-term goal of making it on by default while keeping legacy tests/tools stable.

I can start with Phase 1 now (SB path passes false; `CommandObjectDisassemble `passes false by default; new `--rich` flag), then follow up with a small PR to flip the default and adjust affected tests.

UltimateForce21 committed:
Added `--rich` option for disassembler annotations and updated SBFrame path**

* Added a new `--rich` (`-R`) command-line option to `CommandObjectDisassemble` to enable rich disassembly annotations for the current invocation.

* Plumbed a new `enable_rich_annotations` flag through:
  * `Disassembler::Disassemble` overloads
  * `Disassembler::PrintInstructions`
  * `Instruction::Dump`
  * `StackFrame::Disassemble`

* Updated `StackFrame::Disassemble` to take an optional `bool enable_rich_annotations` (default `false`) so the SB API can request annotated output without CLI involvement.

* Ensured annotations are only added when `enable_rich_annotations` is `true`; preserved caching for the non-rich path.

* Modified `Options.td` to define the new `--rich` option.

* Added/updated API test `TestRichDisassembler.py` to run `disassemble --rich -f` and check annotated output.

* Kept default behavior unchanged so existing scripts and IDE integrations are unaffected.

UltimateForce21 committed:
Formatting changes.

UltimateForce21 committed:
Merge branch 'main' into add-disassembler-annotations

UltimateForce21 committed:
Redo Workflow tests

UltimateForce21 commented:
I’ve addressed all the requested changes and fixed the formatting issues.
The implementation is now complete and all tests are passing locally.

@adrian.prantl, when you have time, could you please review the final version and merge if everything looks good?

Thanks!

UltimateForce21 commented:
> Hi @UltimateForce21! Has this PR been superseded by #152887 and can be closed or is there still contents in here we should try to get merged?

Yes this has been superseded by #152887, so it can be closed. Thanks for checking!
